### PR TITLE
Cursor project level style rules

### PR DESCRIPTION
**Pull request by sdavtaker**
- **Removing cursor rules from .gitignore list**
- **Adding cursor rules for codestyle**

## What type of PR is this? (check all applicable)

- [ ] Feature / Major Change / Refactor / Optimization
- [ ] Bug Fix
- [ ] Documentation Update
- [x] IDE configurations

## Description

Added MDC file for Cursor IDE to follow our Code Style. This is a first iteration, we should refine as we learn. 

## Related tickets & Documents

- Approved RFC (for feature/major changes) #
- Related Issue #
- Closes #

### CODE REVIEW

File: .cursor/rules/code-style.mdc [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+   def process_optional(value: Optional[str]) -> str:
+       return value.upper()  # Could fail if value is None
+   ```
+
+### MyPy Configuration Requirements
+
+- All new code must pass MyPy strict mode
+- Use `# type: ignore` comments only when absolutely necessary and document why
+- Prefer `# type: ignore[error-code]` over blanket ignores
+- Run MyPy locally before committing: `mypy src/`
```
Comment (arapulido) [inline .cursor/rules/code-style.mdc:92]:
```suggestion
- Run MyPy locally before committing: `mypy src/ tests/`
```

File: .cursor/rules/code-style.mdc [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+## OS Operations Must Use Adaptors
+
+**CRITICAL**: Never use OS operations directly in `src/` code. Always use adaptors for better testability and maintainability.
+
+### OS Adaptor Requirements
+
+1. **Use existing adaptors in `src/dd_license_attribution/adaptors/`**:
+   ```python
+   # ✅ Correct
+   from dd_license_attribution.adaptors.os import OSAdaptor
```
Comment (arapulido) [inline .cursor/rules/code-style.mdc:134]: We don't have that class (`OSAdaptor`), just functions

File: .cursor/rules/code-style.mdc [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+       def process_file(self, path: str) -> str:
+           if os.path.exists(path):  # Direct OS usage
+               with open(path) as f:  # Direct OS usage
+                   return f.read()
+           return ""
+   ```
+
+2. **Create new adaptors when needed**:
+   ```python
+   # ✅ Correct - new adaptor
```
Comment (arapulido) [inline .cursor/rules/code-style.mdc:158]: Currently, we are not using classes for adaptors

File: .cursor/rules/code-style.mdc [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+   with patch('dd_license_attribution.adaptors.os.OSAdaptor') as mock_os:
+       mock_os.return_value.read_file.return_value = "test content"
+       service = Service()
+       result = service.process_file("test.txt")
+   ```
+
+### Forbidden Direct Imports in `src/`
+
+- ❌ `import os`
+- ❌ `import sys`
```
Comment (arapulido) [inline .cursor/rules/code-style.mdc:193]: we currently import `sys` in the main_cli

File: .cursor/rules/code-style.mdc [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+- ❌ `import subprocess`
+- ❌ `import pathlib` (use adaptors instead)
+- ❌ `import shutil` (use adaptors instead)
+- ❌ `import tempfile` (use adaptors instead)
```
Comment (arapulido) [inline .cursor/rules/code-style.mdc:197]: we currently import `tempfile` in the maincli

Comment (arapulido) [reply .cursor/rules/code-style.mdc:197]: where does this list of forbidden imports come from?

File: .cursor/rules/code-style.mdc [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+
+## Contract Tests for External Libraries
+
+**REQUIRED**: All external library integrations must have contract tests to ensure compatibility and prevent breaking changes.
+
+### Contract Test Requirements
+
+**Test external API contracts**:
```
Comment (arapulido) [inline .cursor/rules/code-style.mdc:205]: we don't currently have this. where does this come from?

File: .cursor/rules/code-style.mdc [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+   
+   def test_edge_cases(self) -> None:
+       # Test boundary conditions
+       pass
+   ```
+
+3. **Use proper mocking**:
+   ```python
+   # ✅ Correct - mock adaptors
+   @patch('dd_license_attribution.adaptors.os.OSAdaptor')
```
Comment (arapulido) [inline .cursor/rules/code-style.mdc:315]: another example of an non-existing class

File: .cursor/rules/code-style.mdc [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+   # ❌ Incorrect - mock internal implementation
+   @patch('dd_license_attribution.example.ExampleClass._internal_method')
+   def test_with_mocked_internal(self, mock_internal: Mock) -> None:
+       # Don't mock internal implementation details
+   ```
+
+### Pre-commit Checklist
+
+Before committing any code changes:
+- [ ] Run `mypy src/` and ensure no errors
```
Comment (arapulido) [inline .cursor/rules/code-style.mdc:329]:
```suggestion
- [ ] Run `mypy src/ tests/` and ensure no errors
```

**Review by arapulido — CHANGES_REQUESTED** [review]
Thanks for adding this! I made some comments

File: .cursor/rules/code-style.mdc (hunk 3 above)
Comment (sdavtaker) [reply .cursor/rules/code-style.mdc:158]: Yes, but it make sense to group all those methods in a class.

File: .cursor/rules/code-style.mdc (hunk 4 above)
Comment (sdavtaker) [reply .cursor/rules/code-style.mdc:193]: We only using it for checking platform at the entry point, I think its fine to have exceptions, specially in main. I will quote Captain Barbosa about this "The code is more what you'd call 'guidelines' than actual rules" being the code this file :)

**Comment by sdavtaker** [comment]
I committed the 2 suggestions, I don't think we should change the rest at this point. Also, we can always tune this fail as we notice things do not work as expected.

**Review by arapulido — DISMISSED** [review]
sounds good

File: .cursor/rules/code-style.mdc [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+- [ ] Update CHANGELOG.md with any user-facing changes
+- [ ] Run `isort --check-only src/ tests/` and `black --check src/ tests/`
+- [ ] Test that CI MyPy check passes
+- [ ] Test that CI coverage check passes
+- [ ] Test that CI formatting check passes
+
+### Exceptions and Edge Cases
+
+- Use `Any` sparingly and only when the type truly cannot be determined
+- Use `Union[Type1, Type2]` or `Type1 | Type2` (Python 3.10+) for multiple possible types
```
Comment (danisyellis) [inline .cursor/rules/code-style.mdc:346]: do we have a preference for which option is the default when they can both be used?
Maybe whichever one is most prevalent in the codebase?

File: .cursor/rules/code-style.mdc [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+- Use `Literal` for string/enum-like values: `Literal["option1", "option2"]`
+- Use `TypedDict` for dictionary structures with known keys
+- **OS exceptions**: Only allowed in adaptor implementations themselves
+- **Coverage exceptions**: Must be documented and approved by team lead
+
+## Code Formatting Requirements
+
+**CRITICAL**: All code must be formatted using isort and black. CI will block any PR that doesn't comply with these formatting standards.
+
+### Formatting Tools
```
Comment (danisyellis) [inline .cursor/rules/code-style.mdc:356]: perhaps this formatting stuff should go before rules that are running CI. Seems like that would be the last step. (If Cursor looks at/performs actions of these rules in order)

File: .cursor/rules/code-style.mdc [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+   
+   # ❌ Incorrect - mock internal implementation
+   @patch('dd_license_attribution.example.ExampleClass._internal_method')
+   def test_with_mocked_internal(self, mock_internal: Mock) -> None:
+       # Don't mock internal implementation details
+   ```
+
+### Pre-commit Checklist
+
+Before committing any code changes:
```
[the remark this answers is not in the document]
Comment (sdavtaker) [reply .cursor/rules/code-style.mdc:328]: Yes, everything is for proposing code. This is a list of things that it needs to be aware that will be checked by tools at precommit, so it doesn't suggest things that cannot pass the precommit.

File: .cursor/rules/code-style.mdc [hunk 12]
```diff
@@ -0,0 +1,469 @@
...
+- [ ] Run `mypy src/ tests/` and ensure no errors
+- [ ] Verify all new functions have complete type annotations
+- [ ] Check that all variables have appropriate type hints
+- [ ] Ensure no `# type: ignore` comments were added without justification
```
[the remark this answers is not in the document]
Comment (sdavtaker) [reply .cursor/rules/code-style.mdc:332]: Actually means that "if you adding an ignore make a clear comment of why" we don't need to track approvals in code for those, the commit and PR is the approval for it, but we want to be sure that when we look at it in some time we can reevaluate if can be removed or needs to stay longer.

File: .cursor/rules/code-style.mdc (hunk 9 above)
Comment (sdavtaker) [reply .cursor/rules/code-style.mdc:346]: Given we require 3.10+ now, we should limit to the second only, https://peps.python.org/pep-0604/#add-a-new-operator-for-union-type1-type2 is implemented already in 3.10.
Updating in next commit.

File: .cursor/rules/code-style.mdc (hunk 10 above)
Comment (sdavtaker) [reply .cursor/rules/code-style.mdc:356]: It doesn't do sequentiallity (at least not intetionally). But I read that is good to add a closing-sentense/summary phrase at the end with the "Remember: this is important things you need to be aware of"

**Review by danisyellis — APPROVED** [review]
